architecture_records/002-acceleration-framework.md

```diff
@@ -0,0 +1,480 @@
+- parse `configuration_file`, see below, and based on contents, decide the `AccelerationPlugin`'s to promote to `active_plugins`. See [implementaiton of AutoGPTQ LoRA Plugin](#detailed-design) for more description on configuration logic.
+- handle *plugin stacking*, i.e., when we have more than one `active_plugins`, apply their `model_loader` / `augmentation` logic in appropriate succession. 
+    * This is very useful, e.g., loading a LoRA-trainable AutoGPTQ model first, then applying additional optimized fused kernels to further improve training speeds.
+- enforce that `AccelerationFramework.model_loader` witll call `AccelerationPlugin.model_loader` of some `active_plugins` *at most once*.
+    * Prevents potential complication as multiple `active_plugins` could load models in conflicting ways.
+- enforce that `AccelerationFramework.augmentation` would apply `AccelerationPlugin.augmentation` in appropriate succession.
+    * c.f. previous example, where first a LoRA-trainable AutoGPTQ model is loaded, then fused_kernels on applied on top.
+
+```python
+class AccelerationFramework:
+
+    active_plugins: Dict[str, AccelerationPlugin] = dict()
+    plugins_require_custom_loading: List = list()
+
+    def __init__(self, configuration_file: Optional[str]=None):
+
+        with open(configuration_file, "r") as f:
+            contents = yaml.safe_load(f)
+        
+        # pepare the plugin configurations
+        plugin_configs = { k:v for k,v in contents[KEY_PLUGINS].items() }
+
+        for selected_configs, cls in get_relevant_configuration_sections(plugin_configs):
+
+            # then the model is to be installed
+            # get the plugin
+            plugin_name = str(cls.__name__)
+            plugin = cls(selected_configs)
+
+            # check plugin (this is a function that checks if the package requirements of plugin are met)
+            check_plugin_packages(plugin)
+
+            # install plugin
+            self.active_plugins[plugin_name] = plugin
+            if plugin.requires_custom_loading:
+                self.plugins_require_custom_loading.append(plugin_name)
+
+        if len(self.active_plugins) == 0:
+            raise ValueError(
+                "No plugins could be configured. Please check the acceleration "
+                "framework configuration file."
+            )
+
+        assert len(self.plugins_require_custom_loading) <= 1, \
+            f"can load at most 1 plugin with custom model loading, but tried to \'{self.plugins_require_custom_loading}\'."
+
+    def model_loader(self, model_name: str, **kwargs):
+
+        if len(self.plugins_require_custom_loading) == 0:
+            raise NotImplementedError(
+                f"Attempted modeling loading, but none of activated plugins \'{list(self.active_plugins.keys())}\' "
+                "require custom loading."
+            )
+
+        # otherwise there should be exactly 1
```

Comment:
@fabianlim I undertsand the intent, but I feel this logic is making the code a bit complicated :
1. only one plugin can say it requires loading - as contributors add multiple plugins, to avoid errors they need to check other plugins to see if they have custom load implemented. If yes, then they cannot implement load. Error is thrown only at runtime, while creating plugin contributors wont even know the restriction.
2. they will have to go back and change plugin definition in code to fix error 

Instead of this , how about we do following:
1. allow plugins to define whatever they want in classes
2. at runtime, when user passes the config, restrict custom load to 1 by adding a field in yaml
```
plugins :

   peft
   ..... register all

allow_custom_load: plugin name
```

that way user can specify in config at runtime which plugin's load they want to use and the code can just read loader of that pluginClass. Only check will be : if specified plugin does not have a load(), throw warning and say default loader will be used

for augmentation and other callbacks, logic remains same

Reply:
but I guess the reason you might have done it this way is because if a plugib requires custom load, it doesnt make sense to just ignore the load() and apply augmentation?

the above design will only make sense if plugins can have their custom load functions defined, but dont absolutely require them - we can ignore load() and apply augmentation () . If this scenario is not true from use cases we know so far, then your design is fine and might be the only way :)

Reply:
hmm I guess you can have a unit test in the framework that can check all registered plugins and will fail if more than 1 plugin has requires_load property set 

That will prevent any contributions of multiple plugins that require custom load. Mostly I am double checking if this is what you really want to do

Reply:
@raghukiran1224 there is another issue with unsloth, I am discussing with an IP lawyer now on it, but although they use an apache2 license, they have some strange lines of text in 2 of their files asking user not to use those files on 'more than 4 gpu' or else send them money. Current plan is to take those 2 files, and rewrite them.

Reply:
@Ssukriti 

> like if 2 plugins requested - peft and something else say XYZ , both require custom load, we cant use both together and the user has to pick only one?
we cant say - use load from peft, but use augmentation from XYZ without load from XYZ?

Yes correct. The current plan is to keep things simple, and if the user opts to chose 2 plugins that both require model, load, I will raise, and ask user to only pick one.